wellcomeml/ml/bert_semantic_equivalence.py

```diff
@@ -0,0 +1,176 @@
+
+        print(model_name)
+        self.config = BertConfig.from_pretrained(model_name, num_labels=2)
+        self.tokenizer = BertTokenizer.from_pretrained(model_name)
+        self.model = TFBertForSequenceClassification.from_pretrained(
+            model_name, config=self.config, from_pt=from_pt
+        )
+
+    def _prep_dataset_generator(self, X, y):
+        batch_encoding = self.tokenizer.batch_encode_plus(
+            X, max_length=self.max_length, add_special_tokens=True,
+        )
+
+        def gen_train():
+            for i in range(len(X)):
+                features = {k: pad(batch_encoding[k][i], self.max_length)
+                            for k in batch_encoding}
+
+                yield (features, int(y[i]))
+
+        dataset = tf.data.Dataset.from_generator(
+            gen_train,
+            ({"input_ids": tf.int32,
+              "attention_mask": tf.int32,
+              "token_type_ids": tf.int32}, tf.int64),
+            (
+                {
+                    "input_ids": tf.TensorShape([None]),
+                    "attention_mask": tf.TensorShape([None]),
+                    "token_type_ids": tf.TensorShape([None]),
+                },
+                tf.TensorShape([]),
+            ),
+        )
+
+        return dataset
+
+    def fit(self, X, y, random_state=None, epochs=3):
+        """
+        Fits a sentence similarity model
+
+        Args:
+            X: List of 2-uples: [('Sentence 1', 'Sentence 2'), ....]
+            y: list of 0 (not similar) and 1s (similar)
+            random_state: a random state for the train_test_split
+            epochs: number_of epochs
+
+        Returns:
+
+        """
+        # Initialises/downloads model
+        self._initialise_models()
+
+        # Train/val split
+        X_train, X_valid, y_train, y_valid = train_test_split(
+            X, y, test_size=self.test_size, random_state=random_state
+        )
+
+        # Generates tensorflow dataset
+        self.train_dataset = self._prep_dataset_generator(X_train, y_train)
+        self.valid_dataset = self._prep_dataset_generator(X_valid, y_valid)
+
+        # Generates mini-batches and stores in a class variable
+        self.train_dataset = self.train_dataset.shuffle(self.max_length).\
+            batch(self.batch_size).repeat(-1)
+        self.valid_dataset = self.valid_dataset.batch(self.eval_batch_size)
+
+        opt = tf.keras.optimizers.Adam(learning_rate=self.learning_rate,
+                                       epsilon=1e-08)
+
+        loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+
+        # Train and evaluate using tf.keras.Model.fit()
+
+        metric = tf.keras.metrics.SparseCategoricalAccuracy("accuracy")
+
+        self.model.compile(optimizer=opt, loss=loss, metrics=[metric])
+
+        self.train_steps = len(X_train) // self.batch_size
+        self.valid_steps = len(X_valid) // self.eval_batch_size
+
+        self.fit_epoch(epochs=epochs)
```

Comment:
why the need for a separate function?

Reply:
I wanted to have a function that only fits an extra epoch, without having to pass in any data or anything, so I can skip all transformations. I think this could actually be achieved in one function - will look into it 👍

Reply:
This has been fixed in https://github.com/wellcometrust/WellcomeML/pull/43/commits/3d770ba1d5232ec33f3fcf3f2acf7b88bb4536d6